pkg/scheduler/framework/interface.go

```diff
@@ -588,7 +588,10 @@ type Framework interface {
 	// cycle is aborted.
 	// It also returns a PreFilterResult, which may influence what or how many nodes to
 	// evaluate downstream.
-	RunPreFilterPlugins(ctx context.Context, state *CycleState, pod *v1.Pod) (*PreFilterResult, *Status)
+	// The third returns value contains PreFilter plugin that rejected some or all Nodes with PreFilterResult.
+	// But, note that it doesn't contain any plugin when a plugin rejects this Pod with non-success status,
+	// not with PreFilterResult.
+	RunPreFilterPlugins(ctx context.Context, state *CycleState, pod *v1.Pod) (*PreFilterResult, *Status, sets.Set[string])
```

Comment:
Can we make it part of the PreFilterResult? I think also make sense as failed plugins.

Reply:
I think we should define a new type, but perhaps in a follow-up PR as to not block this PR. My argument is here: https://github.com/kubernetes/kubernetes/pull/122251#discussion_r1637720634

Reply:
We shouldn't make it in PreFilterResult as it's an interface between the framework and PreFilter plugin. 

Also, I'm not a fun of the idea defining a new type just for the sake of this function's returning value.

Reply:
Make sense, I always mixed framework interface with plugin interface.

I camp up with a potential risk here about the hint func, as we know, we always check the hint func following the pattern as **before-failed, now-succeeded**, so if we didn't cover the logic in the Prefilter, which may cause the pod stuck in unschedulablePods. 

Take the nodeAffinity for example, we didn't cover the PreFilter in hint func, only care about the Filter logic, so theoretically, we may have problem here. But we don't just because a historical reason here:
```
	// TODO: also check if the original node meets the pod's requestments once preCheck is completely removed.
	// See: https://github.com/kubernetes/kubernetes/issues/110175
```

But before this one, since no plugin registered, so when requeueing, we'll always queued with backoffQ. Is this a valid concern? @sanposhiho

Reply:
> we always check the hint func following the pattern as before-failed, now-succeeded, so if we didn't cover the logic in the Prefilter, which may cause the pod stuck in unschedulablePods.

I believe we don't have a problem with in-tree plugins for this point.
In the first place, we only use PreFilterResult in NodeAffinity and VolumeBinding. VolumeBinding doesn't have QHint yet. And, hence NodeAffinity is actually the only plugin that returns PreFilterResult and already has QHint, but as you also said, QHint looks OK for PreFilterResult failure scenario too.
Actually, even let's say we already implemented that `TODO`, I believe the QHint still should be OK for PreFilterResult failure case too.

Reply:
> I believe we don't have a problem with in-tree plugins for this point.

Yes, this is just a **NOTE** at this time point. And will not conflict with hint func in the future, just we need to pay more attention for PreFilter when writing hint funs.

> even let's say we already implemented that TODO, I believe the QHint still should be OK for PreFilterResult failure case too.

Let's discuss about this, other parts are ok to me. What I mean is if we still follow the **before-failed, now-succeeded** pattern, once a Pod failed by the NodeAffinity PreFilter(specify several nodes), but success with Filter logic, for hint func(the TODO is solved), it will be `before-succeeded, now-succeeded` and skip the requeueing. Unless we also cover the preFilter logic in hint func.

If the analysis is right, this maybe a technical debt, and we should at least update the TODO comments to cover the PreFill logic, or the pod maybe stuck. That's my key point.

Reply:
I don't still see your point. Let me elaborate in case we're not on the same understanding. 

Let's say the Pod has NodeAffinity with `metadata.name` selector for Node1, Node2, and Node3.
We have three scenarios that involves PreFilterResult.

First, if the cluster doesn't have Node1 - 3 yet. 
In this case, PreFilterResult rejects all Nodes, then only the creation of Node(s) with a matching name can make the Pod schedulable, from NodeAffinity PoV.
QHint should return Queue only for the creation event of Node1, 2, or 3. 
This would be correctly handled even after `TODO` is done.

And the second case, if the cluster only has Node1.
In this case, PreFilterResult rejects all Nodes except Node1. But, let's say Node1 is too small and NodeResourceFit rejects Node1. In this case, the Pod goes back to the queue with `unschedPlugin: [NodeAffinity, NodeResourceFit]`.
QHint should return Queue only for the creation event of Node2 or Node3. There's nothing to do with Node1; no need to return Queue for any update to Node1.
This would be correctly handled even after `TODO` is done. 

And, the last case, if actually the cluster already has Node1-Node3.
In this case, NodeAffinity filters out all Nodes edcept Node1-3 with PreFilterResult, and if, for example, NodeResourceFit rejects the three Node at Filter, then the Pod goes back to the queue with `unschedPlugin: [NodeAffinity, NodeResourceFit]`.
But, note that, in this case, the Pod has NodeAffinity in unschedPlugin though, actually NodeAffinity QHint can return Skip for all Node creation/update events.
No more Node creation could make the Pod schedulable and there's nothing to do with Node1-3; no need to return Queue for any update to Node1-3.
This would be correctly handled even after `TODO` is done. 

---

So, I don't see any potential issue even after `TODO` is done.

Reply:
Thanks for the explanation, seems nodeAffinity is not a good example, if nodeAffinity select nodes via label selector in preFill, not specify the node names, then an update event could make the pod stuck, but this is beyond the scope of in-tree plugins.

Back to my original point, consider the prefilter logic in hint func is necessary sometimes. This is just a suggestion not blockers.
/approve

Reply:
> Back to my original point

Yeah, PreFilterResult in in-tree (nodeaffinity, volumebinding) is anyway very simple now. 
But, especially for those who are building complex custom plugins, we should emphasize which part of plugins could be result in `pInfo.UnschedulablePlugins` and thus should be taken care in QHint(s) implementation.